### ISSUE TITLE
FeliCa: Handle card moved out of field better

### ISSUE THREAD

**Issue by micolous** [micolous/metrodroid#41]
## Unsupported card requests

n/a

## Description of the issue

Reading a FeliCa card takes some time.  If the card is moved out of the field, there are strange errors given to the user.

## For card-specific issues

- Name of the card type: Suica / Pasmo / others
- Name of the transit operator: JR East
- Location: Japan

## Steps to reproduce

1. Hold a FeliCa card (ideally Suica/IC compatible) against the phone.
2. Remove it from the field part way through reading.

## Expected behaviour

Something that indicates that the card left the NFC field.

## Actual behaviour

Exceptions within `nfc-felica-lib` are displayed in a dialog

## Workarounds (if any)

Hold the card "properly".

## Log

```
06-15 16:37:00.588 12623-12623/? W/System.err: java.lang.NullPointerException: Attempt to invoke virtual method 'byte[] net.kazzz.felica.lib.FeliCaLib$IDm.getBytes()' on a null object reference
06-15 16:37:00.589 12623-12623/? W/System.err:     at net.kazzz.felica.lib.FeliCaLib$CommandPacket.<init>(FeliCaLib.java:276)
06-15 16:37:00.589 12623-12623/? W/System.err:     at net.kazzz.felica.FeliCaTag.readWithoutEncryption(FeliCaTag.java:268)
06-15 16:37:00.589 12623-12623/? W/System.err:     at au.id.micolous.metrodroid.card.felica.FelicaCard.dumpTag(FelicaCard.java:190)
06-15 16:37:00.590 12623-12623/? W/System.err:     at au.id.micolous.metrodroid.card.Card.dumpTag(Card.java:92)
06-15 16:37:00.590 12623-12623/? W/System.err:     at au.id.micolous.metrodroid.activity.ReadingTagActivity$ReadingTagTask.doInBackground(ReadingTagActivity.java:194)
06-15 16:37:00.590 12623-12623/? W/System.err:     at au.id.micolous.metrodroid.activity.ReadingTagActivity$ReadingTagTask.doInBackground(ReadingTagActivity.java:186)
06-15 16:37:00.590 12623-12623/? W/System.err:     at android.os.AsyncTask$2.call(AsyncTask.java:333)
06-15 16:37:00.590 12623-12623/? W/System.err:     at java.util.concurrent.FutureTask.run(FutureTask.java:266)
06-15 16:37:00.590 12623-12623/? W/System.err:     at android.os.AsyncTask$SerialExecutor$1.run(AsyncTask.java:245)
06-15 16:37:00.590 12623-12623/? W/System.err:     at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1162)
06-15 16:37:00.590 12623-12623/? W/System.err:     at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:636)
06-15 16:37:00.590 12623-12623/? W/System.err:     at java.lang.Thread.run(Thread.java:764)
```

**Comment by micolous** [comment]
**Status as of v2.9.35 (~Aug 2018):** partial reads now throw a more sensible warning for all cards (not NullPointerException), and not just for FeliCa.

That _technically_ addresses the original concern (strange errors), but FeliCa reading in general is still not as reliable as I'd like it. 

I don't know yet whether that's an issue with Metrodroid's reader code, an issue with Android, or an hardware problem.  It's not just impacting Suica, but seems to impact Octopus cards as well.  And probably Kartu Multi-Trip.

It would be nice to be able to _recover_ from this condition -- the `reader-mode` branch should help here, but that's still a ways off.

**Comment by micolous** [comment]
**Status as of what's on `git master` right now:**

I've merged the improvements in #372 using a rebase on another PR (so that's actually "merged").

This contains a near-total rewrite of the FeliCa reading code, which is much faster and more reliable in most cases:

* we don't do as much "Polling" (unless you have a broken/old Octopus card)
* we don't try to read files which would be protected
* there is now a progress indicator

In the case of Suica, card reading time has been reduced from from 6.0 seconds to 4.5 seconds, and it reads successfully about 95% of the time, instead of less than 50%.

I'm going to call this "fixed" on that basis.  The patches will be in v2.9.38, when I cut a release.
